browser/registry.go

```diff
@@ -148,38 +154,136 @@ func (r *remoteRegistry) isRemoteBrowser() (string, bool) {
 	return wsURL, true
 }
 
-// browserRegistry stores browser instances indexed per
-// iteration as identified by VUID-scenario-iterationID.
+// browserRegistry stores a single VU browser instances
+// indexed per iteration.
 type browserRegistry struct {
+	vu k6modules.VU
+
 	mu sync.RWMutex
-	m  map[string]api.Browser
+	m  map[int64]api.Browser
+
+	buildFn browserBuildFunc
+}
+
+type browserBuildFunc func(ctx context.Context) (api.Browser, error)
+
+func newBrowserRegistry(vu k6modules.VU, remote *remoteRegistry, pids *pidRegistry) *browserRegistry {
+	bt := chromium.NewBrowserType(vu)
+	builder := func(ctx context.Context) (api.Browser, error) {
+		var (
+			err                    error
+			b                      api.Browser
+			wsURL, isRemoteBrowser = remote.isRemoteBrowser()
+		)
+
+		if isRemoteBrowser {
+			b, err = bt.Connect(ctx, wsURL)
+			if err != nil {
+				return nil, err //nolint:wrapcheck
+			}
+		} else {
+			var pid int
+			b, pid, err = bt.Launch(ctx)
+			if err != nil {
+				return nil, err //nolint:wrapcheck
+			}
+			pids.registerPid(pid)
+		}
+
+		return b, nil
+	}
+
+	r := &browserRegistry{
+		vu:      vu,
+		m:       make(map[int64]api.Browser),
+		buildFn: builder,
+	}
+
+	iterSubID, eventsCh := vu.Events().Local.Subscribe(
+		k6event.IterStart,
+		k6event.IterEnd,
+	)
+	unsubscribe := func() {
+		vu.Events().Local.Unsubscribe(iterSubID)
+	}
+	go r.handleIterEvents(eventsCh, unsubscribe)
+
+	return r
 }
 
-func newBrowserRegistry() *browserRegistry {
-	return &browserRegistry{
-		m: make(map[string]api.Browser),
+func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsubscribeFn func()) {
+	var (
+		ok    bool
+		data  k6event.IterData
+		ctx   = context.Background()
+		vuCtx = k6ext.WithVU(r.vu.Context(), r.vu)
+	)
+
+	for e := range eventsCh {
+		// If browser module is imported in the test, NewModuleInstance will be called for
+		// every VU. Because on VU init stage we can not distinguish to which scenario it
+		// belongs or access its options (because state is nil), we have to always subscribe
+		// to each VU iter events, including VUs that do not make use of the browser in their
+		// iterations.
+		// Therefore, if we get an event that does not correspond to a browser iteration, then
+		// unsubscribe for the VU events and exit the loop in order to reduce unuseful overhead.
+		if !isBrowserIter(r.vu) {
+			unsubscribeFn()
+			e.Done()
+			return
+		}
+
+		if data, ok = e.Data.(k6event.IterData); !ok {
+			k6ext.Abort(vuCtx, "unexpected iteration event data format: %v", e.Data)
+			e.Done()
+			continue
+		}
+
+		switch e.Type { //nolint:exhaustive
+		case k6event.IterStart:
+			b, err := r.buildFn(ctx)
+			if err != nil {
+				k6ext.Abort(vuCtx, "error building browser on IterStart: %v", err)
+				e.Done()
+				continue
```

Comment:
I believe we have to `continue` here so we keep reading the events and the producer from k6 event system is not blocked. Otherwise the test was getting stuck. Overall this continue should not matter as the abort should make the whole test exit eventually. Another option might be to `unsubscribe` and then return, but I think it's not necessary to complicate it that way.
Maybe we could add a comment in the `continue`? Or what do you think?

Reply:
Aha, i see, that makes it a little tricky.

My preference would be to do some cleanup prior to calling abort in that case. So call `e.Done`, `unsubscribe`, and then `continue`. A comment would also be useful as it's not clear without going down the call stack that `Abort` won't `panic`, but will instead signal to k6 to abort the whole test run.

Reply:
I think we should not unsubscribe here, as we still want to wait for the `Exit` event to do a final cleanup in case some browser was not properly closed by the end of its iteration. I will add a comment explaining the continue.

Reply:
See [1](https://github.com/grafana/xk6-browser/pull/944/commits/a84847afe5637f7f425f569714e2c3cf269adb68#diff-f2c9c90a415d446c8ce22e2d961cf6c20fe8327cbf408087ed7abf1eeb5c256cR251-R253), [2](https://github.com/grafana/xk6-browser/pull/944/commits/a84847afe5637f7f425f569714e2c3cf269adb68#diff-f2c9c90a415d446c8ce22e2d961cf6c20fe8327cbf408087ed7abf1eeb5c256cR239-R241) in [a84847a](https://github.com/grafana/xk6-browser/pull/944/commits/a84847afe5637f7f425f569714e2c3cf269adb68).